test/e2e/dynamic_resource_e2e_test.go

```diff
@@ -3,103 +3,164 @@ package e2e
-			},
+	Context("Given a Namespace with an OwnNamespace OperatorGroup", func() {
+		var (
+			ns         *corev1.Namespace
+			err        error
+			deleteOpts *metav1.DeleteOptions
+		)
+
+		BeforeEach(func() {
+			// Create Namespace
+			ns, err = c.KubernetesInterface().CoreV1().Namespaces().Create(&corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: genName("ns-"),
+				},
+			})
+			Expect(err).NotTo(HaveOccurred())
 		})
-		require.NoError(GinkgoT(), err)
-
-		deleteOpts := &metav1.DeleteOptions{}
-		defer func() {
-			require.NoError(GinkgoT(), c.KubernetesInterface().CoreV1().Namespaces().Delete(ns.GetName(), deleteOpts))
-		}()
-
-		catsrc := &v1alpha1.CatalogSource{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      genName("dynamic-catalog-"),
-				Namespace: ns.GetName(),
-			},
-			Spec: v1alpha1.CatalogSourceSpec{
-				Image:      "quay.io/olmtest/catsrc_dynamic_resources:e2e-test",
-				SourceType: v1alpha1.SourceTypeGrpc,
-			},
-		}
-
-		catsrc, err = crc.OperatorsV1alpha1().CatalogSources(catsrc.GetNamespace()).Create(catsrc)
-		require.NoError(GinkgoT(), err)
-
-		defer func() {
-			require.NoError(GinkgoT(), crc.OperatorsV1alpha1().CatalogSources(catsrc.GetNamespace()).Delete(catsrc.GetName(), deleteOpts))
-		}()
-
-		// Wait for the CatalogSource to be ready
-		catsrc, err = fetchCatalogSource(GinkgoT(), crc, catsrc.GetName(), catsrc.GetNamespace(), catalogSourceRegistryPodSynced)
-		require.NoError(GinkgoT(), err)
-
-		// Generate a Subscription
-		subName := genName("dynamic-resources")
-		cleanupSub := createSubscriptionForCatalog(GinkgoT(), crc, catsrc.GetNamespace(), subName, catsrc.GetName(), "etcd", "singlenamespace-alpha", "", v1alpha1.ApprovalAutomatic)
-		defer cleanupSub()
-
-		sub, err := fetchSubscription(GinkgoT(), crc, catsrc.GetNamespace(), subName, subscriptionHasInstallPlanChecker)
-		require.NoError(GinkgoT(), err)
-
-		// Wait for the expected InstallPlan's execution to either fail or succeed
-		ipName := sub.Status.InstallPlanRef.Name
-		ip, err := waitForInstallPlan(GinkgoT(), crc, ipName, sub.GetNamespace(), buildInstallPlanPhaseCheckFunc(v1alpha1.InstallPlanPhaseFailed, v1alpha1.InstallPlanPhaseComplete))
-		require.NoError(GinkgoT(), err)
-		// Ensure the InstallPlan contains the steps resolved from the bundle image
-		expectedSteps := map[registry.ResourceKey]struct{}{
-			{Name: "my-prometheusrule", Kind: "PrometheusRule"}: {},
-			{Name: "my-servicemonitor", Kind: "ServiceMonitor"}: {},
-		}
-
-		for _, step := range ip.Status.Plan {
-			key := registry.ResourceKey{
-				Name: step.Resource.Name,
-				Kind: step.Resource.Kind,
-			}
-			for expected := range expectedSteps {
-				if strings.HasPrefix(key.Name, expected.Name) && key.Kind == expected.Kind {
-					delete(expectedSteps, expected)
+
+		AfterEach(func() {
+
+			deleteOpts = &metav1.DeleteOptions{}
+			// Delete Namespace
+			err := c.KubernetesInterface().CoreV1().Namespaces().Delete(ns.GetName(), deleteOpts)
```

Comment:
IIRC, there are efforts to clean the namespace before deleting it. I wonder if you want to add them here before deleting the ns. I understand the current work would delete all resources that are in the ns, but may cause deletion to hang if some resources decide to stick around. 

On the worst case, I wonder if adding delediate deletion/gracePeriod=0 would help.

Reply:
@Bowenislandsong , what resources should be cleaned up before deleting a namespace? I was under the impression that [`cleaner.NotifyTestComplete(true)` ](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/test/e2e/dynamic_resource_e2e_test.go#L20)which internally calls `cleanupOLM` deletes OLM CR's.

Reply:
ahh. I see. Then this is good. Thanks for the explanation!